Java/C#/Python prefixed size support

robert-schmidtke opened the pull request:
Hi,

this is a WIP PR that hopefully eventually adds the possibility to prefix the buffer's size to it, just like in C++ with `FinishSizePrefixed` and `GetSizePrefixedRoot`. I have successfully tested the Java part in a separate project of mine, however the C# variant is untested. I would therefore like to ask your advice on how to best add unit tests for the changes, and how to run them on either a Linux or macOS box.

In the `getSizePrefixedRoot` code I generate, I always assume the length is 4 bytes. Having read the documentation, this is intentional, right? Otherwise there would need to be some check (but since `ByteBuffer.capacity()` is an `int` that will not be necessary, I think).

In general I'm of course happy for any feedback, and whether my understanding of the `FlatBufferBuilder`s is correct, and whether my approach of implementation is compatible with the C++ version (a unit test one could add).

The build is succeeding: https://travis-ci.org/robert-schmidtke/flatbuffers/builds/282171784 Having not looked too much into what `make test` actually does, this may not cover all cases. I'll be working simultaneously on adding tests.

Cheers
Robert

robert-schmidtke commented:
`I signed it!`

robert-schmidtke commented:
Both the Java and C# tests succeed as well now, the Java part of which also tests the C++ written, binary size prefixed monster file.

aardappel commented:
You can test under Linux with Mono, see `tests/FlatBuffers.Test/NetTest.sh`.

You may assume size 4, yes. It be nice to use a constant, though, either and existing or new one in `Constants.java` and `FlatBufferConstants.cs`

aardappel commented on src/idl_gen_general.cpp:
```diff
@@ -828,6 +828,33 @@ void GenStruct(StructDef &struct_def, std::string *code_ptr) {
+    code += ps_method_signature + "(ByteBuffer _psbb, " + struct_def.name + " obj) { ";
+    code += "ByteBuffer _bb = _psbb." + FunctionStart('S') + "lice(); ";
+    if (lang_.language == IDLOptions::kCSharp) {
+      code += "_bb.Position = 4; ";
```
constant?

robert-schmidtke replied:
I have added a constant for Java and C# each. The Java one had to be explicitly `public` though, because the other ones are package-private.

aardappel commented on src/idl_gen_general.cpp:
```diff
@@ -828,6 +828,33 @@ void GenStruct(StructDef &struct_def, std::string *code_ptr) {
     code += ", _bb)); }\n";
+
+    // recreate both methods for the prefixed size version
+    std::string ps_method_name = FunctionStart('G') + "etSizePrefixedRootAs" +
```
This code has a lot of overlap with the code above, any ideas to refactor?

robert-schmidtke replied:
I could see something like I've done in lines [1335+](https://github.com/google/flatbuffers/pull/4445/files/68be4420dda47e8d0600bb19691f03be71503a68#diff-d480fbc6a8bf3b7289fe45114ce2889bR1335). However the method bodies would depend on `i`, sort of forcing the for-switch anti-pattern.

robert-schmidtke replied:
I've pushed a version with less overlap in e84452f.

aardappel commented on src/idl_parser.cpp:
```diff
@@ -2204,8 +2204,13 @@ CheckedError Parser::DoParse(const char *source,
-                file_identifier_.length() ? file_identifier_.c_str() : nullptr);
+      if (opts.prefix_size) {
+        builder_.FinishSizePrefixed(Offset<Table>(toff),
+                  file_identifier_.length() ? file_identifier_.c_str() : nullptr);
```
maybe at least pull the file identifier arg out of the if?

aardappel commented on tests/MyGame/Example/Monster.java:
```diff
@@ -14,6 +14,9 @@
   public static Monster getRootAsMonster(ByteBuffer _bb) { return getRootAsMonster(_bb, new Monster()); }
   public static Monster getRootAsMonster(ByteBuffer _bb, Monster obj) { _bb.order(ByteOrder.LITTLE_ENDIAN); return (obj.__assign(_bb.getInt(_bb.position()) + _bb.position(), _bb)); }
+  public static Monster getSizePrefixedRootAsMonster(ByteBuffer _psbb) { return getSizePrefixedRootAsMonster(_psbb, new Monster()); }
+  public static Monster getSizePrefixedRootAsMonster(ByteBuffer _psbb, Monster obj) { ByteBuffer _bb = _psbb.slice(); _bb.position(4); return getRootAsMonster(_bb, obj); }
```
Not sure why we're creating a new `ByteBuffer` here, ideally this refers to the existing one?

robert-schmidtke replied:
There is a new buffer created, however `slice()` uses the data of the other buffer, there is no copying involved, only position, limit etc. are reset. The same happens with the `Slice()` version I have added to the C# `ByteBuffer`. My reasoning behind this was that I wanted to avoid advancing the passed-in `_psbb` to avoid side-effects for the user. One possibility would be to increment the position by 4, do the regular `getRootAsMonster`, and then decrement it again. However I was worried about concurrent access to the buffer. Any thoughts on this? Would the +4, `getRootAsMonster`, -4 version be okay with you?

aardappel replied:
I'm aware we're not copying the buffer, but I generally like to avoid object allocations where possible.
I guess the + then - trick would be ok, but really we shouldn't need to touch the `position` at all if you see how the existing `getRootAsMonster` works, you can just generate a similar method that does `_bb.position() + 4` on the fly.

aardappel commented on tests/monster_test_ps.fbs:
```diff
@@ -0,0 +1,10 @@
+// similar schema, except to be parsed with --prefix-size
+
+include "monster_test.fbs";
```
I'd prefer it if we keep the test simple without needing yet another schema. You could simply generate a size-prefixed buffer in code, and then see if you can access it correctly.

robert-schmidtke replied:
I understand that, I have thought about it similarly. However, I'm not quite sure how else I may check the interoperability between C++ and Java (in this case), if not for a generated file. One could add another dedicated test that checks for interoperability in all supported languages? This test would first generate the regular and the size-prefixed version, and then invoke a small test for each language. This way, the new schema would not need to be versioned in git.

aardappel replied:
I'm not sure that an intra-language test is that important, since all we'd be checking is if a language uses the right size (32 bits)..

robert-schmidtke commented:
I have removed the explicit tests for the size prefixed version, and added inline tests. This brought to my attention, that it might be useful to introduce `SizePrefixedMonsterBufferHasIdentifier` methods that do not rely on the fact, that the user has advanced the provided buffer by 4 bytes themselves?

aardappel commented:
I was thinking, that rather than introducing all these new generated methods (which few people will use, and may confuse), why don't we simply have non-generated functions that can either advance a size prefixed `ByteBuffer`, or slice into a new `ByteBuffer`? There is nothing type specific about these functions. You'd then simply call `getRootAsMonster(bb.FromSizePrefixed())` or whatever.

robert-schmidtke commented:
Sorry, with the last push I added the Python code as well, because I had merged the things on my end already. Guess I'll close the Python PR and extend this one to include the Python changes as well.

aardappel commented:
@robert-schmidtke do you have plans to finish this PR, and/or can @yazdan (see https://github.com/google/flatbuffers/issues/4632) help you get it merged?

robert-schmidtke commented:
Hi @aardappel, thanks for coming back to me on this. I'm using my version of the branch successfully (mostly with Java though) already. I'd like to see the changes merged if the project can benefit from that. Please let me know how I can help with this PR!

robert-schmidtke commented:
I have merged the current master and regenerated the source files. The Java, C# any Python tests all succeed on my Mac.

aardappel commented on net/FlatBuffers/ByteBuffer.cs:
```diff
@@ -40,21 +40,25 @@ public class ByteBuffer
 
         public int Position {
-            get { return _pos; }
+            get { return _pos + _off; }
```
We didn't need this `_off` in Java, why here? It is both not great for performance (since `Position` is used a lot), and makes it harder to tell if this code is correct, i.e. if `_off` is taken into account in all the right places.

robert-schmidtke replied:
There is probably a way to implement this a little more efficiently using `unsafe` C# if that does not violate your guidelines. Note that the Java `ByteBuffer` implementation does something similar for each `get`, e.g. in `HeapByteBuffer`:

```Java
protected int ix(int i) {
    return i + offset;
}

public byte get(int i) {
    return hb[ix(checkIndex(i))];
}

final int checkIndex(int i) {
    if ((i < 0) || (i >= limit))
        throw new IndexOutOfBoundsException();
    return i;
}
```

robert-schmidtke replied:
My final note on this will be that I'll vote for a similar abstraction in C# as in Java:

```csharp
using System;

namespace FlatBuffers
{
    public class ByteBuffer
    {
        private readonly byte[] _buffer;
        private readonly int _cap; // calculated capacity
        private int _pos;          // Must track start of the buffer.
        private readonly int _off; // parent's position if sliced

        public int Length { get { return _cap; } }

        public byte Get(int pos) { return _buffer[_off + pos]; }

        public void Put(int pos, byte b) { _buffer[_off + pos] = b; }

        public ByteBuffer(byte[] buffer) : this(buffer, 0) { }

        public ByteBuffer(byte[] buffer, int pos) : this(buffer, pos, 0) { }

        private ByteBuffer(byte[] buffer, int pos, int off)
        {
            _buffer = buffer;
            _pos = pos;
            _off = off;
            _cap = buffer.Length - _off;
        }

        public int Position {
            get { return _pos; }
            set { _pos = value; }
        }

        public void Reset()
        {
            _pos = 0;
        }

        // Create a new ByteBuffer on the same underlying data.
        // The new ByteBuffer's position will be 0, but starting
        // from this ByteBuffer's current position.
        public ByteBuffer Slice()
        {
            return new ByteBuffer(_buffer, 0, Position);
        }

        // ...
}
```

aardappel commented on src/flatc.cpp:
```diff
@@ -115,6 +115,7 @@ std::string FlatCompiler::GetUsageString(const char *program_name) const {
     "  --no-ts-reexport   Don't re-export imported dependencies for TypeScript.\n"
     "  --reflect-types    Add minimal type reflection to code generation.\n"
     "  --reflect-names    Add minimal type/name reflection.\n"
+    "  --prefix-size      Prefix size field to generated buffers.\n"
```
Maybe note that this only applies to parsed JSON?

robert-schmidtke replied:
it is applicable for parsed JSON and schema serialization, right?

aardappel replied:
True.. just thinking of a way this can be clearer. Maybe `wire format (-b)` instead of `buffers` ?

robert-schmidtke replied:
That makes sense, I'll add that.

aardappel commented on src/idl_parser.cpp:
```diff
@@ -2355,9 +2355,16 @@ CheckedError Parser::DoParse(const char *source, const char **include_paths,
-                                               ? file_identifier_.c_str()
-                                               : nullptr);
+      if (opts.prefix_size) {
+        builder_.FinishSizePrefixed(Offset<Table>(toff),
```
`builder_.Finish(toff, file_identifier_.length() ? file_identifier_.c_str() : nullptr, opts.prefix_size)`, no if-then needed :)

aardappel commented on src/idl_parser.cpp:
```diff
@@ -2464,7 +2471,11 @@ void Parser::Serialize() {
       root_struct_def_ ? root_struct_def_->serialized_location : 0);
-  builder_.Finish(schema_offset, reflection::SchemaIdentifier());
+  if (opts.prefix_size) {
+    builder_.FinishSizePrefixed(schema_offset, reflection::SchemaIdentifier());
```
same here

robert-schmidtke commented:
Okay so I've added quite a few changes to `ByteBuffer`: it now hides the internal byte array (which I think is an appropriate design choice, similar to Java), and it keeps track of the offset into a parent's buffer, in case it was created via `Slice()`. Furthermore I've added more tests that check the slices, as well as size prefix operations. The implementation is fairly close to what Java does internally, I don't think there is much of a performance drawback by carrying a parent's offset around. I'm not a C# expert, but my guess is that the JIT will probably be smart and figure out that the offset is zero in most of the cases, right?

So anyway, let me know what you think. I realize there is some potential API breaking in here.

aardappel commented:
Sorry to be a pain, but still not happy with the addition of `off_`. It is adding additions etc to pretty much all core functions that need to be very fast. Maybe the JIT is smart enough, maybe it is not, I'd rather not take this risk that this is going to make every single FlatBuffers access slower just for the benefit of more efficient slices. The core functionality simply is more imporant than slices.

robert-schmidtke commented:
So I managed to encapsulate the slicing behavior in a subclass of `ByteBuffer`, such that 'pure' `ByteBuffer`s do not suffer from the overhead of carrying the `_off`. Is that more acceptable for you?

aardappel commented on net/FlatBuffers/ByteBuffer.cs:
```diff
@@ -38,12 +40,12 @@ namespace FlatBuffers
-        public int Length { get { return _buffer.Length; } }
+        public virtual int Length { get { return _buffer.Length; } }
 
-        public byte[] Data { get { return _buffer; } }
```
Why is `Data` removed? was this unused?

robert-schmidtke replied:
`Data` was used, mainly for length checks and memory copies. I thought that fully hiding the underlying buffer was slightly better design, especially when dealing with slices. Thus, the buffer's content is modified/read using the `Put*` and `Get*` methods.

aardappel commented on net/FlatBuffers/ByteBuffer.cs:
```diff
@@ -63,11 +65,65 @@ public void Reset()
+            _buffer = newBuffer;
+        }
+
+        public virtual byte[] ToArray(int pos, int len)
```
some methods here are `virtual` and others not.. any reason? I don't think we intend this class to be inherited from, certainly not if it affects performance for basic functions like GetInt etc, which it may.

aardappel commented on net/FlatBuffers/ByteBuffer.cs:
```diff
@@ -63,11 +65,65 @@ public void Reset()
+            return ToArray(Position, Length - Position);
+        }
+
+        public byte[] ToFullArray()
```
what's the use of this one? Why would anyone ever want the unused part of the buffer copied?

robert-schmidtke replied:
It's mainly for testing to check that the padding is correct. This was previously done by checking against `Data` directly.

aardappel commented on net/FlatBuffers/ByteBuffer.cs:
```diff
@@ -63,11 +65,65 @@ public void Reset()
+
+        // Increases the size of the ByteBuffer, and copies the old data towards
+        // the end of the new buffer.
+        public virtual void GrowFront(int newSize)
```
Where did this function come from / where is it called from?

robert-schmidtke replied:
By hiding `Data`, I moved growing the buffer from `FlatBufferBuilder` to `ByteBuffer`. This is a 1:1 copy of that method. It's called from `FlatBufferBuilder::GrowBuffer`.

aardappel commented on net/FlatBuffers/ByteBufferSlice.cs:
```diff
@@ -0,0 +1,232 @@
+            return base.GetUshort(_off + index);
+        }
+
+        public override int GetInt(int index)
```
This function is called a LOT, so it potentially becoming slower by being virtual I am afraid is not acceptable.

robert-schmidtke replied:
Then I frankly do not see another way to support slices, which imho is the cleanest way of representing size-prefixed buffers. Note that using `ByteBuffer` in Java has exactly the same design, even for non-slices. The last resort would be do create a new `ByteBuffer` with a separate underlying `byte[]` array and copy all the previous contents to it, minus the size prefix. That would be a one-time overhead for creating the slice, and about 100% memory overhead, unless the caller discards all references to the size-prefixed buffer, and lets the GC handle that.

robert-schmidtke replied:
Just FYI, I have added a small benchmark:

```csharp
[FlatBuffersTestMethod]
public void BenchmarkFlatBuffer()
{
    Random rnd = new Random(0);
    ByteBuffer bb = new ByteBuffer(File.ReadAllBytes(@"Resources/monsterdata_test.mon"));
    // bb = bb.Slice(); // toggle for ByteBufferSlice performance
    int maxIndex = bb.Length - 4;
    int sum = 0;

    Stopwatch sw = Stopwatch.StartNew();
    for (int i = 0; i < 1000000; ++i)
    {
        sum += bb.GetInt(rnd.Next(maxIndex));
    }
    sw.Stop();

    // write sum to avoid loop optimization
    Console.WriteLine("Regular: {0}, {1} GetInt ops/s", sum, 1000000.0 / sw.Elapsed.TotalMilliseconds);
}
```

I ran it on my Mac ten times with the regular `ByteBuffer` and with the `ByteBufferSlice`. Numbers are `GetInt` operations per second.

Version | Safe | Unsafe
--- | --- | ---
`ByteBuffer` | 33,017 | 44,017
`ByteBufferSlice` | 32,400 | 42,877

So there's a 1.9% slow-down for the safe implementation, and a 2.6% slow-down for the unsafe implementation.

aardappel replied:
You say this is on your Mac.. does that mean these timings are from Mono, or the MS implementation?

If this is using the MS JIT, then indeed you have a good point that performance is acceptable for `_off`. Though part of this may be that the C# ByteBuffer is already quite slow.

I think we can do better though. If you look at how getting the root works in C# generated code:

`  public static Monster GetRootAsMonster(ByteBuffer _bb, Monster obj) { return (obj.__assign(_bb.GetInt(_bb.Position) + _bb.Position, _bb)); }`

This tells me, that to create a root from a size-prefixed buffer, all we need to do is pass it a `ByteBuffer` whose `_pos` has been moved forward by 4 bytes. All buffer accesses from there on are absolute, using the `bb_pos` stored in these table accessor structs as the starting point, so no need to keep adding `_off` at each access. And likely more robust since we can't forget to add `_off` somewhere.

Am I missing something?

robert-schmidtke replied:
Hi, the timings are from Mono:
```bash
Mono JIT compiler version 5.4.1.7 (2017-06/e66d9abbb27 Wed Oct 25 12:10:41 EDT 2017)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
```

Compiled with mcs:
```bash
Mono C# compiler version 5.4.1.0
```

I'm not sure there is an alternative on Mac? Should I get my hands on a Windows machine and benchmark it there as well?

I was thinking to separate `_pos` and `_off`, because `_pos` can be set externally, and therefore the size prefix could be restored (accidentally, e.g. by `Reset()`). For the use case where `bb_pos` is used as offset (when interacting with the `ByteBuffer` through `Monster`) you're correct, but since the `ByteBuffer` may be used rather freely I figured it'd be better to logically separate them.

I have added your suggestion, and these are the results:

Version | Safe | Unsafe
--- | --- | ---
original `ByteBuffer` (current `master`) | 33,876 | 44,436
`ByteBuffer` using `+ _off` | 33,790 (-0.3%) | 43,837 (-1.3%)
`ByteBuffer` using `virtual` | 33,017 (-2.5%) | 44,017 (-1.0%)
`ByteBufferSlice` using `override` | 32,400 (-4.4%) | 42,877 (-3.5%)
`ByteBuffer` using advanced `_pos` | 32,263 (-4.8%) | 45,593 (+2.6%)

I don't really know why this performs so poorly, I guess it's simply not the exact same system load today than it was last time. Your approach would certainly save a lot of code.

aardappel replied:
Supposedly parts of MS's implementation are portable now and run on Linux, and I presume OS X as well, but haven't looked into it. It is entirely possible that Mono is not quite as optimized.

I'd prefer it if we use just a single variable, for efficency (possibly), especially since size-prefixing is such a minor feature compared to the base functionality. To screw this up, someone would have to load a size-prefixed buffer, slice it to obtain a new one, then call Reset on it, and somehow not understand that will get you your size prefixing back? That sounds unlikely to me.

I'm not sure if we're getting the correct results here. You're only calling `bb.GetInt` which doesn't even touch `_pos`! It directly uses its argument into `_buffer`.

aardappel commented:
Otherwise looks good to me now!

robert-schmidtke commented:
Should I squash the commits into one, or will you do that upon merging eventually?

aardappel commented:
No, I'll squash it on merge.
Just waiting for CI. I presume you are "done" with this PR?

robert-schmidtke commented:
Unless you have any more comments/requested changes, I don't think I have anything to add.

aardappel commented:
Thanks for sticking with it :)
